### ISSUE TITLE
An asynchronous socket operation is already in progress using this SocketAsyncEventArgs instance.

### ISSUE THREAD

**Issue by jp7677** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#221]
Hi there,

We have a pretty straight forward ASP .Net Core application with EF Core that does quite some we api calls with database actions concurrently. We are also do make heavy use of a PaginaList class like described here https://docs.microsoft.com/en-us/aspnet/core/data/ef-mvc/sort-filter-page#add-paging-functionality-to-the-index-method.

Usage looks like:
```
using (var dataContext = dataContextFactory.Create())
{
  var query = dataContext.Entity
  .Include(x => x.SomeSubEntity)
  .ThenInclude(x => x.SomeField)
  .ThenInclude(x => x.SomeField2)
  .Include(x => x.SomeSubEntity2)
  .ThenInclude(x => x.SomeField)
  .ThenInclude(x => x.SomeField2)
  .ThenInclude(x => x.SomeField3)
  .AsQueryable();

  query = AddListFiltering(query, filter);
  query = AddListSorting(query, filter);

  return await PaginatedList<Entity>.CreateAsync(query, filter.Page ?? DefaultPage, filter.PageSize ?? DefaultPageSize);
}

```

Unfortunately we are seeing the following exception:

### The issue
```
Exception message: An exception occurred in the database while iterating the results of a query.
Logger: Microsoft.EntityFrameworkCore.Query.Internal.MySqlQueryCompilationContextFactory
Stack trace:
System.InvalidOperationException: An asynchronous socket operation is already in progress using this SocketAsyncEventArgs instance.
   at System.Net.Sockets.SocketAsyncEventArgs.StartConfiguring()
   at System.Net.Sockets.SocketAsyncEventArgs.SetBufferInternal(Byte[] buffer, Int32 offset, Int32 count)
   at MySql.Data.Protocol.Serialization.SocketByteHandler.<DoWriteBytesAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at MySql.Data.ValueTaskExtensions.ContinueWith[T,TResult](ValueTask`1 valueTask, Func`2 continuation)
   at MySql.Data.Protocol.Serialization.ProtocolUtility.WritePacketAsync(IByteHandler byteHandler, Int32 sequenceNumber, ArraySegment`1 contents, IOBehavior ioBehavior)
   at MySql.Data.Protocol.Serialization.ProtocolUtility.WritePayloadAsync(IByteHandler byteHandler, Func`1 getNextSequenceNumber, ArraySegment`1 payload, IOBehavior ioBehavior)
   at MySql.Data.Serialization.MySqlSession.TryAsync[TArg](Func`3 func, TArg arg, IOBehavior ioBehavior, CancellationToken cancellationToken)
   at MySql.Data.Serialization.MySqlSession.SendAsync(PayloadData payload, IOBehavior ioBehavior, CancellationToken cancellationToken)
   at MySql.Data.MySqlClient.CommandExecutors.TextCommandExecutor.<ExecuteReaderAsync>d__3.MoveNext()
```

Once in a while we are also seeing this exception instead:
```
 MySql.Data.MySqlClient.MySqlException: There is already an open DataReader associated with this Connection which must be closed first.
   at MySql.Data.MySqlClient.MySqlCommand.VerifyValid()
   at MySql.Data.MySqlClient.MySqlCommand.<ExecuteNonQueryAsync>d__46.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySql.Data.MySqlClient.MySqlConnection.<BeginDbTransactionAsync>d__6.MoveNext()
```

Not sure though if this is the same issue.
The stacktrace goes further back into our code, but the location there seems to vary. Often it already goes wrong on the `CountAsync` in the PaginatedList class.

### Further technical details

MySQL version: 10.1.21-MariaDB-1~jessie - mariadb.org binary distribution
Operating system: Ubuntu 16.04/Docker 17.03.0-ce, build 60ccb22
Pomelo.EntityFrameworkCore.MySql version: 1.1.1-prerelease-10020
Framework: .Net Core 1.1

Please let me know if I can assist any further.

**Comment by jp7677** [comment]
Seems like the root cause has been found and fixed, cool!
Could you please trigger a new preview/myget build? I would love to test if this issue is really resolved.

**Comment by Kagamine** [comment]
10021

**Comment by jp7677** [comment]
Sorry, I still got this exception once in a while, though it seems less often.

**Comment by Kagamine** [comment]
@bgrainger Could you follow up this issue?

**Comment by caleblloyd** [comment]
@jp7677 is your code open source so that we can look?

Are you firing off multiple tasks on the same DbContext at the same time, then awaiting them later?  Best practice is to await each async call within the DbContext as it is made, by design a MySQL Connection can only service one query at a time.

**Comment by caleblloyd** [comment]
@jp7677 also are you sure that you upgraded to [1.1.1-prerelease-10021](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/1.1.1-prerelease-10021)?

It is currently showing `0` downloads on NuGet, maybe you are still referencing an older prerelease:

![image](https://cloud.githubusercontent.com/assets/2414837/24147977/23b43ad4-0e13-11e7-9f0d-00c6e64663bb.png)

**Comment by Kagamine** [comment]
@caleblloyd I confirm I've downloaded the 10021. There were about 70+ failed requests during 31K+ requests in my web app with pomelo 10020, and there were about 10+ failed requests during 31K+ requests with 10021.

**Comment by caleblloyd** [comment]
I took a look at `MySqlConnector` and cannot find another clear reason why this would occur.  As far as I know, `MySqlDataReader` and `MySqlTransaction` are the only classes that call code when they are disposed, and both of these are properly disposed when the connection is closed.

Right now my hunch is that there is a race condition with the locking in [SynchronizedMySqlDataReader](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/master/src/Pomelo.EntityFrameworkCore.MySql/Storage/SynchronizedMySqlDataReader.cs).  I'd need to spend some time enhancing the performance tests to try and recreate this issue to track down a potential race condition.  We rely on the core Entity Framework library to use up a reader by buffering the results into objects, so tracking exactly where synchronization ends is difficult.

In my opinion, a better fix would be to buffer all result sets upon calling `ExecuteReader` so that the connection is always open to service new calls, no matter what the core Entity Framework library does with the result set.  I've opened https://github.com/mysql-net/MySqlConnector/issues/202 to discuss an option to buffer Result Sets in `MySqlConnector`, which already has all of the logic to buffer result sets written into it, it just needs an option to enable result set buffering.

**Comment by caleblloyd** [comment]
Buffered result sets are now supported in MySqlConnector, and I've added that functionality in #226. 
 Hopefully that clears up this issue.

**Comment by jp7677** [comment]
Hi there, testing with MySqlConnector 0.15.1 and Pomelo.EntityFrameworkCore.MySql 1.1.1-prerelease-10024 I'm still seeing various errors when performing several web api requests with database actions concurrently. Stacktraces do look like:

```
An asynchronous socket operation is already in progress using this SocketAsyncEventArgs instance.
   at System.Net.Sockets.SocketAsyncEventArgs.StartConfiguring()
   at System.Net.Sockets.SocketAsyncEventArgs.SetBufferInternal(Byte[] buffer, Int32 offset, Int32 count)
   at MySql.Data.Protocol.Serialization.SocketByteHandler.<DoWriteBytesAsync>d__4.MoveNext()
```
```
System.InvalidOperationException: Packet received out-of-order. Expected 2; got 1.
   at MySql.Data.MySqlClient.MySqlDataReader.ActivateResultSet(ResultSet resultSet)
   at MySql.Data.MySqlClient.MySqlDataReader.<ReadFirstResultSetAsync>d__62.MoveNext()
```
```
System.InvalidOperationException: Packet received out-of-order. Expected 2; got 102.
   at MySql.Data.MySqlClient.MySqlDataReader.ActivateResultSet(ResultSet resultSet)
   at MySql.Data.MySqlClient.MySqlDataReader.<ReadFirstResultSetAsync>d__62.MoveNext()
```
```
System.Net.Sockets.SocketException: Broken pipe
   at MySql.Data.SocketAwaitable.GetResult()
   at MySql.Data.Protocol.Serialization.SocketByteHandler.<DoWriteBytesAsync>d__4.MoveNext()
```

I don't think that we misuse the library. We are using Identityserver, therefore we need to register a dbcontext like 
```
services.AddScoped(serviceProvider =>
  Container.GetInstance<ISsoDataContextFactory>().Create());
```
For our own code we do inject only a factory and create a different dbcontext type only when really doing something with it like
```
using (var dataContext = dataContextFactory.Create())
{
 // some code
}
```
The SapientGuardian.EntityFrameworkCore.MySql provider is just happy with the way we are doing things.

@caleblloyd Unfortunately I can't share my project :(


PS: If anyone tests with a web api api, disable caching in Chrome is quite important in my case.

**Comment by jp7677** [comment]
Just to be sure, I had also added BufferResultSets=true to my connection string.

**Comment by caleblloyd** [comment]
> with database actions concurrently

Are you executing multiple Entity Framework queries at the same time in the same DbContext?  Right now there is a lock at each command, but Transactions are not locked so they could potentially try to execute while a command is in progress.

Do you pass a connection or a connection string in the `UseMySql` function? Can you post the `UseMySql()` syntax that you use in these 2 functions:

- `ISsoDataContextFactory>().Create()`
- `dataContextFactory.Create()`

> Unfortunately I can't share my project

It'd be extremely helpful if you could pull out just the `Program.cs`, `Startup.cs`, and a controller action that exhibits this bug.  That would at least get me started on how to recreate...

**Comment by jp7677** [comment]
Please take a look at https://github.com/jp7677/hellocoreclr/tree/mariadb.
This sample project of mine describes some of the mechanics I described, though without IdentityServer.

I have created a branch for using MariaDB with your EF provider instead of Sqlite. The second web page of that app (http://localhost:5000/#/greetings) fires two web api request concurrently to the server. Even there I could reproduce some exceptions after hitting several time refresh in Chrome. Though this time it is:

```
System.FormatException: Expected to read 0x00 but got 0x03
   at MySql.Data.ByteArrayReader.ReadByte(Byte value)
   at MySql.Data.Serialization.OkPayload.Create(PayloadData payload)
   at MySql.Data.Serialization.MySqlSession.<TryPingAsync>d__28.MoveNext()
```
I hope this helps to get to the bottom of this...

**Comment by jp7677** [comment]
Ok, just played with my sample app, once in a while I'm seeing the other exception as well, especially with my last commit.

**Comment by caleblloyd** [comment]
@jp7677 thanks for posting the test project, it was very helpful in tracking down the bug!

It turns out that this bug has been lurking in our code for some time now, and is only exposed when a DB Context is configured with a generic, `UseMySql<TContext>()`.  The buggy code instantiated a MySqlConnection object and would re-use the same connection for every single request.  This is why you were seeing conflicts, every single DbContext instantiated was using the same DbConnection, which is a huge issue.

I'm guessing that most of our users configure without generics `UseMySql()`, and this codepath was not affected.

I've got #228 open right now.  I've already tested it against your `hellocoreclr` repo and the fix works.  I'll ship a prerelease in a few minutes.

Thanks for working with us to help track this down!
